File: python/nodes/read_m3u_file.py
from typing import Callable, TypedDict

from ambler import Next, Nextable
from state import State


class ReadEdges(TypedDict):
    on_success: Nextable


class ReadUtils(TypedDict):
    read_text_file: Callable[[str], str]


def _read_text_file(path: str) -> str:
    with open(path, 'r') as f:
        return f.read()


_default_utils: ReadUtils = {
    'read_text_file': _read_text_file,
}


def read_m3u_file(edges: ReadEdges, utils: ReadUtils = _default_utils) -> Nextable:
    async def _node(state: State) -> Next[State]:
        if not state['m3u_file_path']:
            raise ValueError("M3U file path is not defined.")

        content = utils['read_text_file'](state['m3u_file_path'])
        urls = [line.strip() for line in content.splitlines()
                if line.strip() and not line.strip().startswith('#')]

        print(f"Found {len(urls)} URLs in {state['m3u_file_path']}")
        return Next(edges['on_success'], {**state, 'urls': urls})

    return _node

File: typescript/nodes/read_m3u_file.ts
import { Next, Nextable } from "../ambler.ts";

export interface ReadM3UFileState {
  m3uFilePath: string | null;
  urls: string[];
}

type ReadEdges<S> = { onSuccess: Nextable<S> };
type ReadUtils = { readTextFile: (path: string) => Promise<string> };

const defaultUtils: ReadUtils = {
  readTextFile: (path) => Deno.readTextFile(path),
};

export function readM3UFile<S extends ReadM3UFileState>(
  edges: ReadEdges<S>,
  utils: ReadUtils = defaultUtils
): Nextable<S> {
  return async (state: S): Promise<Next<S>> => {
    if (!state.m3uFilePath) {
      throw new Error("M3U file path is not defined.");
    }

    const content = await utils.readTextFile(state.m3uFilePath);
    const urls = content.split("\n")
      .map((line) => line.trim())
      .filter((line) => line.length > 0 && !line.startsWith("#"));

    console.log(`Found ${urls.length} URLs in ${state.m3uFilePath}`);
    return new Next(edges.onSuccess, { ...state, urls });
  };
}
